diff --git a/src/CommandArray.as b/src/CommandArray.as
index a0cdfa9..9b7b1e0 100644
--- a/src/CommandArray.as
+++ b/src/CommandArray.as
@@ -36,7 +36,7 @@ package
 
         public function CommandArray(data:String)
         {
-            array    = data.split("\x00");
+            array    = data.split("\x01");
             length   = array.length;
             position = 0;
         }
diff --git a/src/FlashCanvas.as b/src/FlashCanvas.as
index 25730cb..cf7e8f8 100644
--- a/src/FlashCanvas.as
+++ b/src/FlashCanvas.as
@@ -84,12 +84,7 @@ package
             //
             // @see http://bugs.adobe.com/jira/browse/FP-383
 
-            var objectId:String = ExternalInterface.objectID;
-            if (objectId == null)
-            {
-                objectId = loaderInfo.parameters.id;
-                resize(stage.stageWidth, stage.stageHeight);
-            }
+            var objectId:String = loaderInfo.parameters.id;
 
             // Remove the prefix "external" from objectID
             canvasId = objectId.slice(8);
diff --git a/src/externs.js b/src/externs.js
index d10a3eb..0a61437 100644
--- a/src/externs.js
+++ b/src/externs.js
@@ -13,5 +13,6 @@ G_vmlCanvasManager.init;
 G_vmlCanvasManager.init_;
 G_vmlCanvasManager.initElement;
 swf.CallFunction;
+swf.executeCommand;
 swf.resize;
 swf.saveImage;
diff --git a/src/flashcanvas.js b/src/flashcanvas.js
index 93b6499..e2cfab3 100644
--- a/src/flashcanvas.js
+++ b/src/flashcanvas.js
@@ -11,7 +11,7 @@
 //   http://dev.w3.org/html5/spec/the-canvas-element.html
 
 // If the browser is IE and does not support HTML5 Canvas
-if (window["ActiveXObject"] && !window["CanvasRenderingContext2D"]) {
+if (!window["FlashCanvas"]) {
 
 (function(window, document, undefined) {
 
@@ -27,10 +27,10 @@ var CANVAS_PATTERN              = "CanvasPattern";
 var FLASH_CANVAS                = "FlashCanvas";
 var G_VML_CANVAS_MANAGER        = "G_vmlCanvasManager";
 var OBJECT_ID_PREFIX            = "external";
-var ON_FOCUS                    = "onfocus";
-var ON_PROPERTY_CHANGE          = "onpropertychange";
-var ON_READY_STATE_CHANGE       = "onreadystatechange";
-var ON_UNLOAD                   = "onunload";
+var DOM_CONTENT_LOADED          = "DOMContentLoaded";
+var FOCUS                       = "focus";
+var PROPERTY_CHANGE             = "propertychange";
+var UNLOAD                      = "unload";
 
 var config   = window[FLASH_CANVAS + "Options"] || {};
 var BASE_URL = config["swfPath"] || getScriptUrl().replace(/[^\/]+$/, "");
@@ -145,7 +145,7 @@ var CanvasRenderingContext2D = function(canvas, swf) {
     this._swf = swf;
 
     // unique ID of canvas
-    this._canvasId = swf.id.slice(8);
+    this._canvasId = swf.name.slice(8);
 
     // initialize drawing states
     this._initialize();
@@ -536,14 +536,14 @@ CanvasRenderingContext2D.prototype = {
      */
 
     _setFontStyles: function() {
-        var queue = this._queue;
+        var queue = this._queue, style;
         if (this._font !== this.font) {
             try {
                 var span = spans[this._canvasId];
                 span.style.font = this._font = this.font;
 
-                var style = span.currentStyle;
-                var fontSize = span.offsetHeight;
+                style = getComputedStyle(span, null);
+                var fontSize = parseFloat(style.height);
                 var font = [style.fontStyle, style.fontWeight, fontSize, style.fontFamily].join(" ");
                 queue.push(properties.font, font);
             } catch(e) {
@@ -559,8 +559,9 @@ CanvasRenderingContext2D.prototype = {
             this._textBaseline = this.textBaseline;
             queue.push(properties.textBaseline, this._textBaseline);
         }
-        if (this._direction !== this.canvas.currentStyle.direction) {
-            this._direction = this.canvas.currentStyle.direction;
+        style = getComputedStyle(this.canvas, null);
+        if (this._direction !== style.direction) {
+            this._direction = style.direction;
             queue.push(properties.direction, this._direction);
         }
     },
@@ -594,7 +595,7 @@ CanvasRenderingContext2D.prototype = {
 
         // Replace space characters with tab characters because innerText
         // removes trailing white spaces.
-        span.innerText = text.replace(/[ \n\f\r]/g, "\t");
+        span.textContent = text.replace(/[ \n\f\r]/g, "\t");
 
         return new TextMetrics(span.offsetWidth);
     },
@@ -774,10 +775,7 @@ CanvasRenderingContext2D.prototype = {
         // execute commands
         var commands = this._flush();
         if (commands.length > 0) {
-            return eval(this._swf.CallFunction(
-                '<invoke name="executeCommand" returntype="javascript"><arguments><string>'
-                + commands.join("&#0;") + "</string></arguments></invoke>"
-            ));
+            return this._swf.executeCommand(commands.join("\1"));
         }
     },
 
@@ -860,15 +858,12 @@ var DOMExceptionNames = {
 /*
  * Event handlers
  */
+function onDomContentLoaded() {
+    document.removeEventListener(DOM_CONTENT_LOADED, onDomContentLoaded, false);
 
-function onReadyStateChange() {
-    if (document.readyState === "complete") {
-        document.detachEvent(ON_READY_STATE_CHANGE, onReadyStateChange);
-
-        var canvases = document.getElementsByTagName(CANVAS);
-        for (var i = 0, n = canvases.length; i < n; ++i) {
-            FlashCanvas.initElement(canvases[i]);
-        }
+    var canvases = document.getElementsByTagName(CANVAS);
+    for (var i = 0, n = canvases.length; i < n; ++i) {
+        FlashCanvas.initElement(canvases[i]);
     }
 }
 
@@ -900,10 +895,10 @@ function onPropertyChange() {
 }
 
 function onUnload() {
-    window.detachEvent(ON_UNLOAD, onUnload);
+    window.removeEventListener(UNLOAD, onUnload, false);
 
     for (var canvasId in canvases) {
-        var canvas = canvases[canvasId], swf = canvas.firstChild, prop;
+        var canvas = canvases[canvasId], swf = canvas, prop;
 
         // clean up the references of swf.executeCommand and swf.resize
         for (prop in swf) {
@@ -920,8 +915,8 @@ function onUnload() {
         }
 
         // remove event listeners
-        swf.detachEvent(ON_FOCUS, onFocus);
-        canvas.detachEvent(ON_PROPERTY_CHANGE, onPropertyChange);
+        swf.removeEventListener(FOCUS, onFocus, false);
+        canvas.removeEventListener(PROPERTY_CHANGE, onPropertyChange, false);
     }
 
     // delete exported symbols
@@ -939,9 +934,9 @@ function onUnload() {
 var FlashCanvas = {
     initElement: function(canvas) {
         // Check whether the initialization is required or not.
-        if (canvas.getContext) {
-            return canvas;
-        }
+//      if (canvas.getContext) {
+//          return canvas;
+//      }
 
         // initialize lock
         var canvasId        = getUniqueId();
@@ -955,35 +950,39 @@ var FlashCanvas = {
         setCanvasSize(canvas);
 
         // embed swf and SPAN element
-        canvas.innerHTML =
-            '<object classid="clsid:d27cdb6e-ae6d-11cf-96b8-444553540000"' +
-            ' codebase="' + location.protocol + '//fpdownload.macromedia.com/pub/shockwave/cabs/flash/swflash.cab#version=9,0,0,0"' +
-            ' width="100%" height="100%" id="' + objectId + '">' +
+        var div = document.createElement("div"), swf;
+        div.innerHTML = '<object type="application/x-shockwave-flash"' +
+            ' data="' + SWF_URL + '"' +
+            ' width="' + canvas.width + '"' +
+            ' height="' + canvas.height + '"' +
+            ' id="' + canvas.id + '"' +
+            ' name="' + objectId + '">' +
             '<param name="allowScriptAccess" value="always">' +
             '<param name="flashvars" value="id=' + objectId + '">' +
+            '<param name="movie" value="' + SWF_URL + '">' +
             '<param name="wmode" value="transparent">' +
             '</object>' +
-            '<span style="margin:0;padding:0;border:0;display:inline-block;position:static;height:1em;overflow:visible;white-space:nowrap">' +
+            '<span style="margin:0;padding:0;border:0;display:inline-block;position:absolute;top:-9999px;height:1em;overflow:visible;white-space:nowrap">' +
             '</span>';
+        canvas.parentNode.replaceChild(div, canvas);
 
-        canvases[canvasId] = canvas;
-        var swf            = canvas.firstChild;
-        spans[canvasId]    = canvas.lastChild;
+        canvases[canvasId] = canvas = swf = div.firstChild;
+        spans[canvasId]    = div.lastChild;
 
         // Check whether the canvas element is in the DOM tree
-        var documentContains = document.body.contains;
-        if (documentContains(canvas)) {
-            // Load swf file immediately
-            swf["movie"] = SWF_URL;
-        } else {
-            // Wait until the element is added to the DOM tree
-            var intervalId = setInterval(function() {
-                if (documentContains(canvas)) {
-                    clearInterval(intervalId);
-                    swf["movie"] = SWF_URL;
-                }
-            }, 0);
-        }
+//      var documentContains = document.body.contains;
+//      if (documentContains(canvas)) {
+//          // Load swf file immediately
+//          swf["movie"] = SWF_URL;
+//      } else {
+//          // Wait until the element is added to the DOM tree
+//          var intervalId = setInterval(function() {
+//              if (documentContains(canvas)) {
+//                  clearInterval(intervalId);
+//                  swf["movie"] = SWF_URL;
+//              }
+//          }, 0);
+//      }
 
         // If the browser is IE6 or in quirks mode
         if (document.compatMode === "BackCompat" || !window.XMLHttpRequest) {
@@ -1009,13 +1008,13 @@ var FlashCanvas = {
         };
 
         // add event listener
-        swf.attachEvent(ON_FOCUS, onFocus);
+        swf.addEventListener(FOCUS, onFocus, false);
 
         return canvas;
     },
 
     saveImage: function(canvas) {
-        var swf = canvas.firstChild;
+        var swf = canvas;
         swf.saveImage();
     },
 
@@ -1039,7 +1038,7 @@ var FlashCanvas = {
         // If Flash becomes ready
         if (url === undefined) {
             canvas = canvases[canvasId];
-            swf    = canvas.firstChild;
+            swf    = canvas;
 
             // Set the width and height attributes of the canvas element.
             setCanvasSize(canvas);
@@ -1059,7 +1058,7 @@ var FlashCanvas = {
             swf.resize(width, height);
 
             // Add event listener
-            canvas.attachEvent(ON_PROPERTY_CHANGE, onPropertyChange);
+            canvas.addEventListener(PROPERTY_CHANGE, onPropertyChange, false);
 
             // ExternalInterface is now ready for use
             isReady[canvasId] = true;
@@ -1146,22 +1145,22 @@ function setCanvasSize(canvas) {
 document.createElement(CANVAS);
 
 // setup default CSS
-document.createStyleSheet().cssText =
-    CANVAS + "{display:inline-block;overflow:hidden;width:300px;height:150px}";
+//document.createStyleSheet().cssText =
+//    CANVAS + "{display:inline-block;overflow:hidden;width:300px;height:150px}";
 
 // initialize canvas elements
 if (document.readyState === "complete") {
-    onReadyStateChange();
+    onDomContentLoaded();
 } else {
-    document.attachEvent(ON_READY_STATE_CHANGE, onReadyStateChange);
+    document.addEventListener(DOM_CONTENT_LOADED, onDomContentLoaded, false);
 }
 
 // prevent IE6 memory leaks
-window.attachEvent(ON_UNLOAD, onUnload);
+window.addEventListener(UNLOAD, onUnload, false);
 
 // preload SWF file if it's in the same domain
 if (SWF_URL.indexOf(location.protocol + "//" + location.host + "/") === 0) {
-    var req = new ActiveXObject("Microsoft.XMLHTTP");
+    var req = new XMLHttpRequest;
     req.open("GET", SWF_URL, false);
     req.send(NULL);
 }
